src/libsyntax/ext/expand.rs

```diff
@@ -282,6 +282,22 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
                     let expansion = self.expand_invoc(invoc, ext);
                     self.collect_invocations(expansion, &[])
                 } else if let InvocationKind::Attr { attr: None, traits, item } = invoc.kind {
+                    if let Annotatable::Item(ref item) = item {
```

Comment:
Could we also error on `Annotatable::TraitItem` and `Annotatable::ImplItem` here and then also get rid of the other corresponding error in `src/libsyntax_ext/deriving/generic/mod.rs`? e.g.
```rust
let derive_allowed = match item {
    Annotatable::Item(ref item) => match item.node {
        ast::ItemKind::Struct(..) | ast::ItemKind::Enum(..) | ast::ItemKind::Union(..) => true,
        _ => false,
    _ => false,
};
if !derive_allowed {
    // error
}
```